src/charm.py

```diff
@@ -33,83 +31,59 @@ def __init__(self, *args):
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+        except CharmConfigInvalidError as exc:
+            self.model.unit.status = ops.BlockedStatus(exc.msg)
+            return
+        self.smtp = smtp.SmtpProvides(self)
+        self.framework.observe(
+            self.on[LEGACY_RELATION_NAME].relation_created, self._on_relation_created
+        )
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent):
-        """Define and start a workload using the Pebble API.
-
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
-
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
-
-        Args:
-            event: event triggering the handler.
-        """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
+        self.framework.observe(self.on.update_status, self._on_update_status)
+
+    def _on_relation_created(self, _) -> None:
+        """Handle a change to the smtp relation."""
+        # A new charm will be instantiated hence, the information will be fetched again.
+        # The relation databags are rewritten in case there are changes.
+        self._update_relations()
+
+    def _on_update_status(self, _) -> None:
+        """Handle the update status event."""
+        # A new charm will be instantiated hence, the information will be fetched again.
+        # The relation databags are rewritten in case there are changes.
+        self._update_relations()
+
+    def _on_config_changed(self, _) -> None:
+        """Handle changes in configuration."""
+        self.unit.status = ops.MaintenanceStatus("Configuring charm")
+        self._update_relations()
         self.unit.status = ops.ActiveStatus()
 
-    def _on_config_changed(self, event: ops.ConfigChangedEvent):
-        """Handle changed configuration.
-
-        Change this example to suit your needs. If you don't need to handle config, you can remove
-        this method.
+    def _update_relations(self) -> None:
+        """Update all SMTP data for the existing relations."""
+        if not self.model.unit.is_leader():
+            return
+        for relation in self.smtp.relations:
+            self.smtp.update_relation_data(relation, self.get_smtp_data())
 
-        Learn more about config at https://juju.is/docs/sdk/config
+    def get_smtp_data(self) -> smtp.SmtpRelationData:
```

Comment:
What is the reason for this being a public method please? Could it be private?

Reply:
Indeed. Changed